cmd/pyroscope/command/agent.go

```diff
@@ -0,0 +1,54 @@
+package command
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pyroscope-io/pyroscope/pkg/cli"
+	"github.com/pyroscope-io/pyroscope/pkg/config"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func newAgentCmd(cfg *config.Agent) *cobra.Command {
+	agentCmd := &cobra.Command{
+		Use:   "agent [flags]",
+		Short: "starts pyroscope agent.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if cfg.Config != "" {
+				// Use config file from the flag.
+				viper.SetConfigFile(cfg.Config)
+
+				// If a config file is found, read it in.
+				if err := viper.ReadInConfig(); err == nil {
+					fmt.Fprintln(os.Stderr, "Using config file:", viper.ConfigFileUsed())
+				}
+
+				if err := viper.Unmarshal(cfg); err != nil {
+					fmt.Fprintln(os.Stderr, "Unable to unmarshal:", err)
+				}
+			}
+
+			err := cli.StartAgent(cfg)
+			if err != nil {
+				cmd.SilenceUsage = true
+			}
```

Comment:
Not sure if it required indeed: `SilenceUsage` suppresses usage when an error occurs so that it can be set to `true` at command initialisation.

Reply:
I think this just makes it so that it doesn't print usage twice or something like that. @Skemba can you please clarify? And I guess a comment here would be great because it's not obvious.

Reply:
I will add comment for this as well, but it was added because we don't want usage to be printed when an error occurs running the subcommand, but we still want it printed when parsing flags and similar fails.

Reply:
> we don't want usage to be printed when an error occurs running the subcommand, but we still want it printed when parsing flags and similar fails.

Aha, I see. I should say that we can achieve exactly this behaviour by setting some additional handlers:

```golang

func newRootCmd() *cobra.Command {
	var version bool
	rootCmd := &cobra.Command{
		Use:           "pyroscope [flags] <subcommand>",
		SilenceUsage:  true,
		SilenceErrors: true,
		Run: func(cmd *cobra.Command, args []string) {
			if version {
				printVersion(cmd, args)
				return
			}
			_ = usageFn(cmd)
		},
	}
	rootCmd.SetUsageFunc(usageFn)
	rootCmd.SetFlagErrorFunc(flagParseErrFn)
	rootCmd.Flags().BoolVarP(&version, "version", "v", false, "")
	return rootCmd
}

func usageFn(cmd *cobra.Command) error {
	cmd.Println(gradientBanner())
	cmd.Println(DefaultUsageFunc(cmd.Flags(), cmd))
	return nil
}

func flagParseErrFn(cmd *cobra.Command, err error) error {
	_ = usageFn(cmd)
	cmd.Println("Error:", err)
	return nil
}
```

```Golang

func newVersionCmd() *cobra.Command {
	return &cobra.Command{
		Use:    "version",
		Hidden: true,
		Run:    printVersion,
	}
}

func printVersion(_ *cobra.Command, _ []string) {
	fmt.Println(gradientBanner())
	fmt.Println(build.Summary())
}

```

----

Note that the snippet also affects how `version` command/version works: now it is printed on:
 - `pyroscope -v`
 - `pyroscope -version`
 - `pyroscope --version`
 - `pyroscope version`

`Version` member of `config.Config` can be removed, I think.

----

We also need to make sure command descriptions are okay. Now it looks as the following:
```
  agent       starts pyroscope agent.
  completion  generate the autocompletion script for the specified shell
  connect     connects to an existing process and profiles it
  exec        starts a new process from arguments and profiles it
  help        Help about any command
  server      starts pyroscope server. This is the database + web-based user interface
```

`completion` and `help` specify the action imperatively, meanwhile other just describe. Please check capitals and punctuation marks.

----

> Although I'm not sure if I should make it even more specific as we might want usage to still be printed for example if exec.Cli failes with "no arguments passed", but not if any other error happens?

We could also use cobra's validation.

----

Another thing is that we log some messages at start time which makes the output a bit inconsistent :/ I think we could omit this.

> INFO[2021-08-19T12:39:50.540138] scope/pkg/exec/cli.go:80 to disable logging from pyroscope, pass -no-logging argument to pyroscope exec 
Error: on macOS you're required to run the agent with sudo

Reply:
> Aha, I see. I should say that we can achieve exactly this behaviour by setting some additional handlers:

That's true, just don't know if we only want usage printed when there's a flag error or is there some other case we want this to cover as well? "We could also use cobra's validation" -> for example this is a good idea if we care about flags+arguments, just wondering if there is something else too?

> Note that the snippet also affects how version command/version works: now it is printed on:

Not sure what is this version part in reference to? 

> completion and help specify the action imperatively, meanwhile other just describe. Please check capitals and punctuation marks.

these are the default python ones, while the others are ours for which I have copied out what text was there previously. How do we want it to be?

> Another thing is that we log some messages at start time which makes the output a bit inconsistent :/ I think we could omit this.

not sure what do you mean exactly exactly? Remove the logs that happen before error checking?

Reply:
0. Usage. I believe we should print out usage message only when subcommand or flag is not valid. Argument value most likely does not have anything to do with CLI.

1. Version. Right now `pyroscope version` and `pyroscope --version` return `Error: unknown command "version" for "pyroscope"`. I'd suggest introducing a separate subcommand. And optionally a flag – just for convenience.

2. Help. What caught my eye is that descriptions for subcommands are written in different fashions:
```
  agent       starts pyroscope agent.
  help        Help about any command
```
I think we should rely on some naming convention. I'd suggest to use cobra's one:
```
  agent       Start pyroscope agent
  help        Help about any command
```

3. Log messages at start. As you can see this log message is not quite appropriate here. Perhaps we should use logger only after user input is validated.